File: Classes/DrawView.m
//
//  DrawView.m
//  DrawView
//
//  Created by Frank Michael on 4/8/14.
//  Copyright (c) 2014 Frank Michael Sanchez. All rights reserved.
//

#import "DrawView.h"

@interface DrawView () {
    NSMutableArray *paths;
    UIBezierPath *bezierPath;
    CAShapeLayer *animateLayer;
    BOOL isAnimating;
    BOOL isDrawingExisting;
    UIBezierPath *signLine;
}
- (IBAction)undoDrawing:(id)sender;
@end

@implementation DrawView

#pragma mark - Init
- (id)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        // Initialization code
        [self setupUI];
    }
    return self;
}
- (id)initWithCoder:(NSCoder *)aDecoder{
    self = [super initWithCoder:aDecoder];
    if (self){
        [self setupUI];
    }
    return self;
}
#pragma mark - UI Configuration
- (void)setupUI{
    // Array of all the paths the user will draw.
    paths = [NSMutableArray new];
    // Default colors for drawing.
    self.backgroundColor = [UIColor whiteColor];
    _strokeColor = [UIColor blackColor];
    _canEdit = YES;
}
- (void)setStrokeColor:(UIColor *)strokeColor{
    _strokeColor = strokeColor;
}
#pragma mark - View Drawing
- (void)drawRect:(CGRect)rect{
    // Drawing code
    if (!isAnimating){
        [_strokeColor setStroke];
        if (!isDrawingExisting){
            // Need to merge all the paths into a single path.
            for (UIBezierPath *path in paths){
                [path strokeWithBlendMode:kCGBlendModeNormal alpha:1.0];
            }
        }else{
            [bezierPath strokeWithBlendMode:kCGBlendModeNormal alpha:1.0];
        }
    }
    
    if (_mode == SignatureMode){
        [[UIColor lightGrayColor] setStroke];
        [signLine strokeWithBlendMode:kCGBlendModeNormal alpha:1.0];
    }
}
- (void)drawPath:(CGPathRef)path{
    isDrawingExisting = YES;
    _canEdit = NO;
    bezierPath = [UIBezierPath new];
    bezierPath.CGPath = path;
    bezierPath.lineCapStyle = kCGLineCapRound;
    bezierPath.lineWidth = _strokeWidth;
    bezierPath.miterLimit = 0.0f;
    // If iPad apply the scale first so the paths bounds is in its final state.
    if ([[[UIDevice currentDevice] model] rangeOfString:@"iPad"].location != NSNotFound){
        [bezierPath setLineWidth:_strokeWidth];
        CGAffineTransform scaleTransform = CGAffineTransformMakeScale(2, 2);
        [bezierPath applyTransform:scaleTransform];
    }
    // Center the drawing within the view.
    CGRect charBounds = bezierPath.bounds;
    CGFloat charX = CGRectGetMidX(charBounds);
    CGFloat charY = CGRectGetMidY(charBounds);
    CGRect cellBounds = self.bounds;
    CGFloat centerX = CGRectGetMidX(cellBounds);
    CGFloat centerY = CGRectGetMidY(cellBounds);
    
    [bezierPath applyTransform:CGAffineTransformMakeTranslation(centerX-charX, centerY-charY)];
    
    [self setNeedsDisplay];
    
    // Debugging bounds view.
    if (_debugBox){
        UIView *blockView = [[UIView alloc] initWithFrame:CGRectMake(bezierPath.bounds.origin.x, bezierPath.bounds.origin.y, bezierPath.bounds.size.width, bezierPath.bounds.size.height)];
        [blockView setBackgroundColor:[UIColor blackColor]];
        [blockView setAlpha:0.5];
        [self addSubview:blockView];
    }
}
- (void)drawBezier:(UIBezierPath *)path{
    [self drawPath:path.CGPath];
}
- (IBAction)undoDrawing:(id)sender{
    [paths removeLastObject];
    [self setNeedsDisplay];
}
- (void)setMode:(DrawingMode)mode{
    _mode = mode;
    if (mode == DrawingModeDefault){
        signLine = nil;
    }else if (mode == SignatureMode){
        signLine = [UIBezierPath new];
        signLine.lineCapStyle = kCGLineCapRound;
        signLine.lineWidth = 3.0f;
        // Draw the X for the line
        [signLine moveToPoint:CGPointMake(20, self.frame.size.height-30)];
        [signLine addLineToPoint:CGPointMake(30, self.frame.size.height-40)];
        [signLine moveToPoint:CGPointMake(30, self.frame.size.height-30)];
        [signLine addLineToPoint:CGPointMake(20, self.frame.size.height-40)];
        // Draw the line for signing on
        [signLine moveToPoint:CGPointMake(20, self.frame.size.height-20)];
        [signLine addLineToPoint:CGPointMake(self.frame.size.width-20, self.frame.size.height-20)];
    }
    [self setNeedsDisplay];
}
- (void)refreshCurrentMode{
    [self setMode:_mode];
}
- (void)clearDrawing{
    bezierPath = nil;
    paths = nil;
    signLine = nil;
    [self setNeedsDisplay];
    [self setupUI];
}
#pragma mark - View Draw Reading
- (UIImage *)imageRepresentation{
    UIGraphicsBeginImageContext(self.bounds.size);
    CGContextRef context = UIGraphicsGetCurrentContext();
    [self.layer renderInContext:context];
    UIImage *viewImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return viewImage;
}
- (UIBezierPath *)bezierPathRepresentation{
    UIBezierPath *singleBezPath = [UIBezierPath new];
    if (paths.count > 0){
        for (UIBezierPath *path in paths){
            [singleBezPath appendPath:path];
        }
    }else{
        singleBezPath = bezierPath;
    }
    return singleBezPath;
}
#pragma mark - Animation
- (void)animatePath{
    UIBezierPath *animatingPath = [UIBezierPath new];
    if (_canEdit){
        for (UIBezierPath *path in paths){
            [animatingPath appendPath:path];
        }
    }else{
        animatingPath = bezierPath;
    }
    // Clear out the existing view.
    isAnimating = YES;
    [self setNeedsDisplay];
    // Create shape layer that stores the path.
    animateLayer = [[CAShapeLayer alloc] init];
    animateLayer.fillColor = nil;
    animateLayer.path = animatingPath.CGPath;
    animateLayer.strokeColor = [_strokeColor CGColor];
    animateLayer.lineWidth = _strokeWidth;
    animateLayer.miterLimit = 0.0f;
    animateLayer.lineCap = @"round";
    // Create animation of path of the stroke end.
    CABasicAnimation *animation = [[CABasicAnimation alloc] init];
    animation.duration = 3.0;
    animation.fromValue = @(0.0f);
    animation.toValue = @(1.0f);
    animation.delegate = self;
    [animateLayer addAnimation:animation forKey:@"strokeEnd"];
    [self.layer addSublayer:animateLayer];
}
#pragma mark - Animation Delegate
- (void)animationDidStop:(CAAnimation *)anim finished:(BOOL)flag{
    isAnimating = NO;
    [animateLayer removeFromSuperlayer];
    animateLayer = nil;
    [self setNeedsDisplay];
}
#pragma mark - Touch Detecting
- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event{
    if (_canEdit){
        bezierPath = [[UIBezierPath alloc] init];
        [bezierPath setLineCapStyle:kCGLineCapRound];
        [bezierPath setLineWidth:_strokeWidth];
        [bezierPath setMiterLimit:0];
        
        UITouch *currentTouch = [[touches allObjects] objectAtIndex:0];
        [bezierPath moveToPoint:[currentTouch locationInView:self]];
        [paths addObject:bezierPath];
    }
}
- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event{
    if (_canEdit){
        UITouch *movedTouch = [[touches allObjects] objectAtIndex:0];
        [bezierPath addLineToPoint:[movedTouch locationInView:self]];
        [self setNeedsDisplay];
    }
}

@end

File: Classes/DrawView.swift
//
//  DrawView.swift
//  Example-Swift
//
//  Created by Frank Michael on 12/24/14.
//  Copyright (c) 2014 Frank Michael Sanchez. All rights reserved.
//

import UIKit

class DrawView: UIView {
    // Drawing Mode enum
    enum DrawingMode {
        case Default
        case Signature
    }
    enum EditingMode {
        case None
        case Editable
    }
    var drawingMode: DrawingMode = .Default {
        didSet {
            self.setNeedsDisplay()
        }
    }
    var editingMode: EditingMode = .Editable
    var strokeColor: UIColor = UIColor.blackColor()
    var strokeWidth: CGFloat = 25.0
    var debugingBox: Bool = false
    var animationDuration: Double = 3.0
    private var paths: [UIBezierPath] = []
    private var bezierPath: UIBezierPath = UIBezierPath()
    private var signLine: UIBezierPath?
    private var existingPath: CGPathRef?
    private var animationLayer: CAShapeLayer?
    private var isAnimating: Bool = false
    // MARK: Init
    override init() {
        super.init()
        self.setupCanvas()
    }
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.setupCanvas()
    }
    required init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self.setupCanvas()
    }
    // MARK: Class setup
    private func setupCanvas() {
        self.backgroundColor = UIColor.whiteColor()
        self.signLine = UIBezierPath()
        self.signLine?.lineCapStyle = kCGLineCapRound
        self.signLine?.lineWidth = 3.0
        // Draw the "x" for the line
        self.signLine?.moveToPoint(CGPointMake(20.0, self.frame.size.height-30.0))
        self.signLine?.addLineToPoint(CGPointMake(30.0, self.frame.size.height-40.0))
        self.signLine?.moveToPoint(CGPointMake(30.0, self.frame.size.height-30.0))
        self.signLine?.addLineToPoint(CGPointMake(20.0, self.frame.size.height-40.0))
        // Draw the actual line for the signature mode
        self.signLine?.moveToPoint(CGPointMake(20.0, self.frame.size.height-20.0))
        self.signLine?.addLineToPoint(CGPointMake(self.frame.size.width-20.0, self.frame.size.height-20.0))
        self.setNeedsDisplay()
    }
    override func drawRect(rect: CGRect) {
        if !isAnimating{
            self.strokeColor.setStroke()
            if let existing = self.existingPath {
                self.bezierPath.strokeWithBlendMode(kCGBlendModeNormal, alpha: 1.0)
            }else{
                for path in self.paths{
                    path.strokeWithBlendMode(kCGBlendModeNormal, alpha: 1.0)
                }
            }
        }
        
        if self.drawingMode == .Signature {
            UIColor.lightGrayColor().setStroke()
            self.signLine?.strokeWithBlendMode(kCGBlendModeNormal, alpha: 1.0)
        }
    }
    // MARK: Public Functions
    func drawPath(path: CGPathRef) {
        self.editingMode = .None
        self.existingPath = path
        
        self.bezierPath = UIBezierPath()
        self.bezierPath.CGPath = self.existingPath!
        self.bezierPath.lineCapStyle = kCGLineCapRound
        self.bezierPath.lineWidth = self.strokeWidth
        self.bezierPath.miterLimit = 0.0
        // If iPad, apply the scale first so the paths bounds is in its final state.
        // Center the drawing within the view
        let pathBounds = self.bezierPath.bounds
        let pathXMid = CGRectGetMidX(pathBounds)
        let pathYMid = CGRectGetMidY(pathBounds)
        let viewBounds = self.bounds
        let viewCenterX = CGRectGetMidX(viewBounds)
        let viewCenterY = CGRectGetMidY(viewBounds)
        
        self.bezierPath.applyTransform(CGAffineTransformMakeTranslation((viewCenterX-pathXMid), (viewCenterY-pathYMid)))
        self.setNeedsDisplay()
        
        if self.debugingBox {
            let blockView = UIView(frame: CGRectMake(self.bezierPath.bounds.origin.x, self.bezierPath.bounds.origin.y, self.bezierPath.bounds.size.width, self.bezierPath.bounds.size.height))
            blockView.backgroundColor = UIColor.blackColor()
            blockView.alpha = 0.5
            self.addSubview(blockView)
        }
    }
    func drawBezierPath(path: UIBezierPath) {
        self.drawPath(path.CGPath)
    }
    func clearCanvas() {
        self.bezierPath = UIBezierPath()
        self.paths.removeAll(keepCapacity: false)
        self.setNeedsDisplay()
        self.setupCanvas()
    }
    func refreshCanvas() {
        self.setupCanvas()
    }
    // MARK Undo support
    func undoLastPath() {
        self.paths.removeLast()
        self.setNeedsDisplay()
    }
    // MARK View Draw Reading
    func imageRepresentation() -> UIImage {
        UIGraphicsBeginImageContext(self.bounds.size)
        let context = UIGraphicsGetCurrentContext()
        self.layer.renderInContext(context)
        let viewImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return viewImage
    }
    func bezierPathRepresentation() -> UIBezierPath {
        var singleBezierPath = UIBezierPath()
        if self.paths.count > 0 {
            for path in paths {
                singleBezierPath.appendPath(path)
            }
        }else{
            singleBezierPath = self.bezierPath
        }
        return singleBezierPath
    }
    // MARK: Animation
    func animateCanvas() {
        var animatingPath = UIBezierPath()
        if self.editingMode == .Editable {
            for path in self.paths {
                animatingPath.appendPath(path)
            }
        }else{
            animatingPath = self.bezierPath
        }
        // Clear out the existing view
        self.isAnimating = true
        self.setNeedsDisplay()
        // Create shapre layer that stores the path
        self.animationLayer = CAShapeLayer()
        self.animationLayer?.fillColor = nil
        self.animationLayer?.path = animatingPath.CGPath
        self.animationLayer?.strokeColor = self.strokeColor.CGColor
        self.animationLayer?.lineWidth = self.strokeWidth
        self.animationLayer?.miterLimit = 0.0
        self.animationLayer?.lineCap = "round"
        // Create animation of path
        let animation = CABasicAnimation()
        animation.duration = self.animationDuration
        animation.fromValue = 0.0
        animation.toValue = 1.0
        animation.delegate = self
        self.animationLayer?.addAnimation(animation, forKey: "strokeEnd")
        self.layer.addSublayer(self.animationLayer?)
    }
    override func animationDidStop(anim: CAAnimation!, finished flag: Bool) {
        self.isAnimating = false
        self.animationLayer?.removeFromSuperlayer()
        self.animationLayer = CAShapeLayer()
        self.setNeedsDisplay()
    }
    // MARK: Touch Tracking
    override func touchesBegan(touches: NSSet, withEvent event: UIEvent) {
        if self.editingMode == .Editable {
            self.bezierPath = UIBezierPath()
            self.bezierPath.lineCapStyle = kCGLineCapRound
            self.bezierPath.lineWidth = self.strokeWidth
            self.bezierPath.miterLimit = 0.0
            
            let touch = touches.allObjects[0] as UITouch
            bezierPath.moveToPoint(touch.locationInView(self))
            self.paths.append(self.bezierPath)
        }
    }
    override func touchesMoved(touches: NSSet, withEvent event: UIEvent) {
        if self.editingMode == .Editable {
            let touch = touches.allObjects[0] as UITouch
            self.bezierPath.addLineToPoint(touch.locationInView(self))
            self.setNeedsDisplay()
        }
    }
}
